File: skills/adobe-agent-code-audit/scripts/engines/commerce/lib/expert.py
"""
Expert Validation & Recommendation Engine
==========================================
Auto-generates enterprise-grade expert recommendations for audit findings
based on Adobe Commerce best practices, severity, and effort level.

Architecture:
- 40 base recommendation templates (domain-specific guidance)
- Category routing rules: default template + issue-type overrides
- Severity-aware prefix: "prioritize:" for CRITICAL/HIGH
- Effort-aware suffix: rollout guidance for High effort
- Special status prefixes: false-positive, optional, low-risk, correct-with-stronger-control
"""

# ─── Base Templates ────────────────────────────────────────────────────────────
# Each template provides domain-specific, actionable expert guidance.
# Keyed by short ID for compact routing rules.

TEMPLATES = {
    "charset_utf8mb4": (
        "Standardize new/custom tables on utf8mb4 with a compatible collation "
        "for the target MySQL version. Plan conversion in batches, verify index "
        "length, and test comparisons/sorting before production rollout."
    ),
    "audit_timestamps": (
        "Add created_at/updated_at only for mutable business records that require "
        "auditability; populate defaults in db_schema.xml/resource model and index "
        "dates only when used in filters or cleanup jobs."
    ),
    "lob_access_patterns": (
        "Review actual access patterns. Split rarely used large/LOB attributes into "
        "extension/detail tables only if row size, cache pressure, or query plans "
        "show measurable impact."
    ),
    "strict_types": (
        "Add declare(strict_types=1) to new or actively touched PHP files, but avoid "
        "bulk-only churn unless CI static analysis is ready. Pair with type hints "
        "and tests to catch compatibility issues."
    ),
    "ddl_validation": (
        "Validate the recommendation with schema intent and production query patterns "
        "before changing DDL. Use declarative schema/data patches and test "
        "rollback/forward deployment behavior."
    ),
    "index_validation": (
        "Add indexes only for real query predicates/orderings. Validate with EXPLAIN "
        "on production-like data to avoid unnecessary write overhead and overly wide indexes."
    ),
    "exception_refine": (
        "Refine catches to expected Magento/domain exceptions first, keep a final "
        "Throwable/Exception fallback only at a boundary, and avoid returning partial "
        "or silent success after failure."
    ),
    "unsigned_ids": (
        "Use unsigned integer types for foreign keys referencing Magento entity IDs. "
        "Plan a data-safe migration if values or FK compatibility could be affected."
    ),
    "deprecated_upgrade": (
        "Prioritize deprecated APIs that block PHP/Adobe Commerce upgrades. Replace "
        "with Laminas/Magento equivalents and verify behavior through automated "
        "regression tests."
    ),
    "fk_lifecycle": (
        "Validate parent-child lifecycle before adding FKs. Add constraints for strong "
        "ownership, but avoid cascade deletes on high-volume commerce tables unless "
        "the data-retention impact is approved."
    ),
    "validate_general": (
        "Recommendation direction looks reasonable. Before implementation, validate "
        "the finding against real execution paths, add a targeted test, and confirm "
        "the fix does not alter checkout/order/customer flows unexpectedly."
    ),
    "logger_injection": (
        "Replace static helper calls with PSR-3 logger injection so logs are testable, "
        "contextual, and channel-aware; include masked IDs and avoid PII/secrets."
    ),
    "service_contracts": (
        "Extract reusable service contracts where there is real reuse, not just cosmetic "
        "refactoring. Preserve backward compatibility for public APIs and DI preferences."
    ),
    "repository_pattern": (
        "Prefer repositories/resource models/collections for domain data. Use direct "
        "connection only for justified bulk operations with typed parameters, "
        "whitelisted identifiers, and covered tests."
    ),
    "reserved_word_rename": (
        "Rename in a backward-compatible migration only when practical; otherwise quote "
        "identifiers consistently and avoid exposing reserved names in new schema."
    ),
    "encrypted_config": (
        "Use encrypted backend models for sensitive config, scope values deliberately, "
        "mask secrets in admin/logs, and move environment-specific values to Cloud "
        "variables/env.php."
    ),
    "cron_stagger": (
        "Stagger heavy jobs, add locking/idempotency, set sane batch sizes, and "
        "monitor duration/failure metrics so jobs do not overlap or block sales operations."
    ),
    "resolver_cache": (
        "Add resolver-level identity/cache metadata only for cacheable reads. Validate "
        "customer/session-specific data is not cached and add integration tests for "
        "cache invalidation."
    ),
    "sql_injection_control": (
        "Do not rely on bind parameters for identifiers such as table/column names; "
        "validate them against an explicit whitelist and quote identifiers via the DB "
        "adapter. Bind all values, cast LIMIT/OFFSET to int, and add integration tests "
        "for malicious inputs."
    ),
    "http_action_interface": (
        "Implement the appropriate HttpGetActionInterface/HttpPostActionInterface and "
        "enforce form key/ACL where needed; this is especially important for admin and "
        "state-changing endpoints."
    ),
    "structured_logging": (
        "Use structured, leveled logs with correlation IDs. Mask PII/secrets, avoid "
        "noisy info logs in hot paths, and configure rotation/retention in Cloud/log "
        "aggregation."
    ),
    "plugin_ordering": (
        "Add explicit sortOrder and disabled attributes, minimize around plugins, and "
        "verify plugin ordering against third-party modules to avoid hidden behavior changes."
    ),
    "observer_thin": (
        "Keep observers thin and idempotent. Move heavy work to services/queues, guard "
        "against repeated dispatch, and document event payload assumptions."
    ),
    "transaction_wrap": (
        "Wrap multi-write operations in beginTransaction/commit/rollBack, ensure "
        "idempotency for retries, and keep external API calls outside the DB "
        "transaction where possible."
    ),
    "batch_load": (
        "Batch-load collections before the loop, select only required fields, preserve "
        "result ordering explicitly, and profile affected pages/cron jobs before and "
        "after the fix."
    ),
    "cache_profiling": (
        "Confirm the change with production-like profiling. Avoid blanket caching for "
        "personalized data; define cache identities/tags and invalidation before enabling."
    ),
    "acl_least_privilege": (
        "Ensure admin routes, WebAPI resources, and controller actions use "
        "least-privilege ACLs. Add negative tests for unauthorized customer/admin/API roles."
    ),
    "cloud_topology": (
        "Validate this against Adobe Commerce Cloud topology: use remote storage/CDN "
        "where appropriate, keep generated files out of Git, and ensure deploy hooks "
        "are idempotent."
    ),
    "utc_timestamp": (
        "Use Magento UTC storage conventions, convert only at presentation boundaries, "
        "and avoid MySQL TIMESTAMP surprises unless timezone conversion is explicitly intended."
    ),
    "sku_unique": (
        "Add a unique constraint only if business rules guarantee one row per SKU; "
        "otherwise add a non-unique index and enforce uniqueness at the correct scope "
        "such as website/store/channel."
    ),
    "false_positive_secrets": (
        "Constants that only hold config XML paths are not secrets; downgrade if no "
        "literal key/token is present. If actual values exist in code/config dumps, "
        "rotate them and move them to encrypted admin config, env.php/Cloud variables, "
        "or a vault with least-privilege ACLs."
    ),
    "security_priority": (
        "Treat as high priority, but verify exploitability and false positives. Apply "
        "least privilege, input validation, output escaping, secret hygiene, and "
        "security regression tests."
    ),
    "di_injection": (
        "Inject dependencies through constructors/factories/proxies instead of "
        "ObjectManager. Watch for heavy services in constructors and use proxies for "
        "expensive or optional dependencies."
    ),
    "controller_thin": (
        "Move orchestration into service classes with interfaces, keep controllers thin, "
        "and add unit/integration coverage around the extracted business rules."
    ),
    "queue_limits": (
        "Set max_messages/batch limits, make consumers idempotent, add retry/dead-letter "
        "handling, and monitor queue lag before increasing concurrency."
    ),
    "escaper_output": (
        "Escape output by context using escaper methods or secure Knockout bindings; "
        "do not double-escape translated phrases and validate rich HTML through an allowlist."
    ),
    "exception_noswallow": (
        "Do not swallow exceptions. Log sanitized context with correlation/order "
        "identifiers, rethrow when state consistency matters, and only suppress "
        "documented non-critical exceptions."
    ),
    "csrf_formkey": (
        "For state-changing POST actions, validate form keys or use "
        "CsrfAwareActionInterface only with a documented safe exception. Never exempt "
        "public write endpoints without compensating auth controls."
    ),
    "upload_security": (
        "Use Magento UploaderFactory or an equivalent service, enforce extension/MIME/size "
        "whitelist, generate a safe filename, store outside pub until validated, and "
        "scan or quarantine uploads before processing."
    ),
    "test_coverage": (
        "Add focused unit/integration/API tests for the corrected behavior and one "
        "regression test for the failure mode; gate high-risk fixes in CI."
    ),
    "business_flow_integrity": (
        "Validate the customization against Adobe Commerce service-layer flow, entity "
        "state machines, events, indexes, emails, and rollback behavior. Add integration "
        "tests for checkout/order/customer side effects and retry/idempotency."
    ),
    "critical_callback_security": (
        "Verify authentication/signature, replay protection, idempotency keys, row locking, "
        "and duplicate/out-of-order handling before changing payment, shipment, refund, or "
        "inventory state."
    ),
    "msi_inventory_api": (
        "Use MSI service contracts/reservation/source-item APIs instead of legacy stock table "
        "writes. Test multi-source, salable qty, backorders, cancellations, refunds, and shipment "
        "deduction behavior."
    ),
    "admin_api_least_privilege": (
        "Apply least-privilege ACL, ownership validation, CSRF/form-key or signed request checks, "
        "and negative API/admin tests for anonymous, wrong customer, and low-privilege admin access."
    ),
}


# ─── Status Prefix Map ────────────────────────────────────────────────────────
# Some issue types get special status prefixes instead of the standard Aligned.

STATUS_PREFIXES = {
    "aligned":           "Aligned",
    "aligned_prioritize": "Aligned; prioritize",
    "optional":          "Optional / style alignment",
    "low_risk":          "Low-risk correction",
    "correct_stronger":  "Correct with stronger control",
    "needs_review":      "Needs review / possible false positive",
}

# Rollout suffix for high-effort items
ROLLOUT_SUFFIX = (
    " Plan rollout in small PRs with rollback notes because "
    "the estimated effort is high."
)


# ─── Category Routing Rules ───────────────────────────────────────────────────
# Each category defines:
#   "default": template_id — used for all issue types unless overridden
#   "overrides": list of (keyword_or_prefix, template_id, optional_status)
#     - keyword matches are case-insensitive substring checks on issue_type
#     - status override replaces the auto-computed status

CATEGORY_RULES = {
    # ── Code Audit Categories ──
    "Exception Handling": {
        "default": "exception_refine",
        "overrides": [
            ("Empty Catch", "exception_noswallow", None),
            ("Debug Output", "validate_general", None),
            ("Missing Finally", "validate_general", None),
            ("Silent Return", "validate_general", None),
            ("Return After Exception", "validate_general", None),
            ("Wrong Log Level", "validate_general", None),
        ],
    },
    "Security": {
        "default": "security_priority",
        "overrides": [
            ("Hardcoded Credentials", "false_positive_secrets", "needs_review"),
            ("Unsafe File Upload", "upload_security", None),
            ("SQL Injection", "sql_injection_control", "correct_stronger"),
            ("Superglobal", "security_priority", None),
            ("Path Traversal", "security_priority", None),
            ("Test Framework", "escaper_output", None),
            ("CSP", "security_priority", None),
        ],
    },
    "Database": {
        "default": "ddl_validation",
        "overrides": [
            ("N+1", "batch_load", None),
            ("Load in Loop", "batch_load", None),
            ("Missing Transaction", "transaction_wrap", None),
            ("Save in Loop", "ddl_validation", None),
            ("Raw SQL", "ddl_validation", None),
            ("Table Name Interpolation", "sql_injection_control", "correct_stronger"),
            ("Missing Index", "audit_timestamps", None),
        ],
    },
    "Caching": {
        "default": "cache_profiling",
        "overrides": [
            ("GraphQL Resolver Without Cache", "resolver_cache", None),
        ],
    },
    "Code Structure": {
        "default": "service_contracts",
        "overrides": [
            ("Missing strict_types", "strict_types", "optional"),
            ("Business Logic in Controller", "controller_thin", None),
        ],
    },
    "Performance": {
        "default": "cache_profiling",
    },
    "Deprecated": {
        "default": "deprecated_upgrade",
    },
    "Logging": {
        "default": "structured_logging",
        "overrides": [
            ("Static Logger", "logger_injection", None),
        ],
    },
    "File Storage": {
        "default": "cron_stagger",
        "overrides": [
            ("S3", "cloud_topology", None),
        ],
    },
    "Reusability": {
        "default": "service_contracts",
    },
    "Test Coverage": {
        "default": "test_coverage",
    },
    "Dependency Injection": {
        "default": "di_injection",
        "overrides": [
            ("Legacy _objectManager", "deprecated_upgrade", None),
        ],
    },
    "Plugin Architecture": {
        "default": "plugin_ordering",
    },
    "Cron Jobs": {
        "default": "cron_stagger",
        "overrides": [
            ("Cron Without Lock", "observer_thin", None),
        ],
    },
    "GraphQL": {
        "default": "resolver_cache",
        "overrides": [
            ("N+1", "batch_load", None),
        ],
    },
    "Queue Processing": {
        "default": "queue_limits",
    },
    "Configuration": {
        "default": "escaper_output",
    },
    "Frontend Templates": {
        "default": "escaper_output",
    },
    "XML Configuration": {
        "default": "encrypted_config",
        "overrides": [
            ("Plugin Missing sortOrder", "plugin_ordering", None),
            ("Plugin Missing disabled", "plugin_ordering", None),
            ("Sandbox", "false_positive_secrets", "needs_review"),
            ("Test URL", "false_positive_secrets", "needs_review"),
            ("Core Class Override", "di_injection", None),
            ("TODO in Cron", "cron_stagger", None),
            ("Sensitive Field Not Encrypted", "encrypted_config", None),
        ],
    },
    "WebAPI & ACL": {
        "default": "acl_least_privilege",
    },
    "DB Schema": {
        "default": "ddl_validation",
        "overrides": [
            ("Status Column Without Index", "lob_access_patterns", None),
            ("Nullable Without Default", "lob_access_patterns", None),
            ("No Indexes on Table", "ddl_validation", None),
            ("created_at", "audit_timestamps", None),
            ("updated_at", "audit_timestamps", None),
        ],
    },
    "Infrastructure": {
        "default": "cloud_topology",
        "overrides": [
            ("Security Header", "observer_thin", None),
            ("X-XSS-Protection", "escaper_output", None),
            ("Rate Limiting", "cloud_topology", None),
        ],
    },
    "Cloud Deployment": {
        "default": "cron_stagger",
        "overrides": [
            ("Missing PHP Extension", "cloud_topology", None),
            ("PHP Extension", "cloud_topology", None),
            ("High Consumer", "queue_limits", None),
            ("Consumer Processes", "queue_limits", None),
        ],
    },
    "PHP Deep Analysis": {
        "default": "validate_general",
        "overrides": [
            ("Direct DB Connection", "repository_pattern", None),
            ("DateTime Without Timezone", "utc_timestamp", None),
            ("Deprecated", "deprecated_upgrade", None),
            ("exit/die", "validate_general", None),
            ("Weak Hashing", "validate_general", None),
        ],
    },
    "Event Observers": {
        "default": "observer_thin",
    },
    "Module Architecture": {
        "default": "service_contracts",
        "overrides": [
            ("Controller Missing HTTP", "http_action_interface", None),
            ("Global Plugin", "service_contracts", None),
        ],
    },
    "Code Metrics": {
        "default": "validate_general",
    },
    "Business Logic Identification": {
        "default": "validate_general",
        "overrides": [
            ("Admin Grid", "deprecated_upgrade", None),
            ("REST API", "acl_least_privilege", None),
        ],
    },
    "Business Customization Review": {
        "default": "business_flow_integrity",
        "overrides": [
            ("Direct Order/Entity State", "business_flow_integrity", "correct_stronger"),
            ("Direct Save", "business_flow_integrity", "correct_stronger"),
            ("Synchronous External API", "critical_callback_security", None),
            ("Hardcoded Business Rule", "validate_general", None),
        ],
    },
    "Critical Commerce Flows": {
        "default": "business_flow_integrity",
        "overrides": [
            ("Webhook/Callback", "critical_callback_security", "correct_stronger"),
            ("Around Plugin", "plugin_ordering", None),
            ("collectTotals", "cache_profiling", None),
        ],
    },
    "MSI Inventory & Source Management": {
        "default": "msi_inventory_api",
    },
    "Admin & Integration Security": {
        "default": "admin_api_least_privilege",
        "overrides": [
            ("Inbound Integration", "critical_callback_security", "correct_stronger"),
            ("Broad WebAPI ACL", "admin_api_least_privilege", "correct_stronger"),
        ],
    },
    "Logical Flow & Cross-Module": {
        "default": "service_contracts",
        "overrides": [
            ("Circular Dependency", "service_contracts", None),
            ("High Coupling", "service_contracts", None),
            ("Central Module", "service_contracts", None),
            ("Duplicated Logic", "service_contracts", None),
            ("Duplicate Class", "service_contracts", None),
            ("Unused Module", "validate_general", "needs_review"),
            ("Missing module.xml Sequence", "validate_general", None),
            ("Cross-Module Event", "observer_thin", None),
            ("Multi-Module Plugin", "plugin_ordering", None),
            ("Cross-Module Analysis Summary", "validate_general", None),
        ],
    },

    # ── DB Audit Categories ──
    "DB: Table Structure": {
        "default": "ddl_validation",
        "overrides": [
            ("No Secondary Indexes", "index_validation", None),
            ("Wide Table", "lob_access_patterns", None),
            ("Extremely Wide", "lob_access_patterns", None),
            ("Moderately Wide", "lob_access_patterns", None),
        ],
    },
    "DB: Index Analysis": {
        "default": "index_validation",
        "overrides": [
            ("Date Column", "audit_timestamps", None),
            ("Wide Composite", "ddl_validation", None),
            ("Redundant Index", "ddl_validation", None),
            ("Excessive Indexes", "audit_timestamps", None),
        ],
    },
    "DB: Column Analysis": {
        "default": "lob_access_patterns",
        "overrides": [
            ("TIMESTAMP", "audit_timestamps", None),
            ("Oversized Boolean", "ddl_validation", None),
            ("Excessive VARCHAR", "ddl_validation", None),
            ("Signed ID", "unsigned_ids", None),
            ("Imprecise Type for Money", "ddl_validation", None),
        ],
    },
    "DB: Foreign Keys": {
        "default": "fk_lifecycle",
    },
    "DB: Naming Conventions": {
        "default": "reserved_word_rename",
        "default_status": "low_risk",
        "overrides": [
            ("Missing Vendor Prefix", "ddl_validation", None),
            ("CamelCase", "ddl_validation", None),
        ],
    },
    "DB: Storage Engine": {
        "default": "fk_lifecycle",
    },
    "DB: Charset & Collation": {
        "default": "charset_utf8mb4",
    },
    "DB: Adobe Commerce Schema": {
        "default": "ddl_validation",
        "overrides": [
            ("Cleanup Tables", "cron_stagger", None),
        ],
    },
    "DB: Data Integrity": {
        "default": "audit_timestamps",
        "overrides": [
            ("Email Column", "ddl_validation", None),
            ("SKU Without UNIQUE", "sku_unique", None),
        ],
    },
    "DB: Performance": {
        "default": "lob_access_patterns",
        "overrides": [
            ("No Clustered Index", "cache_profiling", None),
            ("DB Performance Summary", "cache_profiling", None),
        ],
    },
}


# ─── Expert Recommendation Generator ──────────────────────────────────────────

def get_expert_recommendation(category, issue_type, severity="MEDIUM", effort="Medium"):
    """Generate expert recommendation for a finding based on category, issue type, severity and effort.

    Returns a string like:
        "Aligned; prioritize: <template text>. Plan rollout in small PRs..."
    or  "Needs review / possible false positive: <template text>"
    """
    rules = CATEGORY_RULES.get(category)
    if not rules:
        # Fallback for unknown categories
        return _format_recommendation("validate_general", severity, effort)

    template_id = rules["default"]
    status_override = rules.get("default_status")

    # Check overrides (first keyword match wins)
    for override in rules.get("overrides", []):
        keyword, tmpl_id, s_override = override
        if keyword.lower() in issue_type.lower():
            template_id = tmpl_id
            status_override = s_override
            break

    return _format_recommendation(template_id, severity, effort, status_override)


def _format_recommendation(template_id, severity, effort, status_override=None):
    """Format the final recommendation with appropriate prefix and suffix."""
    base_text = TEMPLATES.get(template_id, TEMPLATES["validate_general"])

    # Determine status prefix
    if status_override:
        prefix = STATUS_PREFIXES.get(status_override, "Aligned")
    elif severity in ("CRITICAL", "HIGH"):
        prefix = "Aligned; prioritize"
    else:
        prefix = "Aligned"

    result = f"{prefix}: {base_text}"

    # Add rollout suffix for high-effort items
    if effort in ("High", "Very High"):
        result += ROLLOUT_SUFFIX

    return result


def get_executive_summary_expert(row_label, severity_context=None):
    """Generate expert recommendation for Executive Summary rows."""
    label = str(row_label).strip()

    if label in ("Generated:", "Project Root:", "Tool:", "Version:"):
        return "Metadata only; no remediation required."

    if "Total Findings" in label:
        return ("Use as a triage count, not a defect count; de-duplicate "
                "repeated patterns before sizing remediation.")

    if "SEVERITY BREAKDOWN" in label or label == "Severity":
        return ("Prioritize by exploitability/customer impact, not severity "
                "label alone; review possible false positives.")

    if label in ("CRITICAL", "HIGH"):
        return ("Prioritize by exploitability/customer impact, not severity "
                "label alone; review possible false positives.")

    return "Review supporting detailed sheets before actioning this summary item."

File: skills/bmad-dept-code-audit-agent/scripts/engines/aem/lib/expert.ts
/**
 * Expert Validation & Recommendation Engine — AEM
 * ================================================
 * Auto-generates enterprise-grade expert recommendations for AEM audit findings
 * based on Adobe Experience Manager best practices, severity, and effort level.
 *
 * Architecture:
 * - 35+ base recommendation templates (AEM-specific guidance)
 * - Category routing rules: default template + issue-type overrides
 * - Severity-aware prefix: "prioritize:" for CRITICAL/HIGH
 * - Effort-aware suffix: rollout guidance for High effort
 * - Special status prefixes: false-positive, optional, low-risk
 */

// ─── Base Templates ────────────────────────────────────────────────────────────
const TEMPLATES: Record<string, string> = {
  resource_resolver:
    "Always close ResourceResolvers obtained from ResourceResolverFactory in a " +
    "try-with-resources block. Leaked resolvers hold JCR sessions open, exhausting " +
    "the session pool and causing cascading failures under load.",
  service_user:
    "Replace admin session usage with service-user-based ResourceResolvers mapped " +
    "via Sling Service User Mapping. Define least-privilege ACLs and verify the " +
    "mapping exists in the repo-init scripts.",
  sling_model_best_practice:
    "Use @Model with adaptables=Resource (preferred) or SlingHttpServletRequest. " +
    "Avoid WCMUsePojo; use injector-specific annotations (@ChildResource, @ValueMapValue) " +
    "over generic @Inject for clarity and fail-fast behavior.",
  osgi_config_separation:
    "Separate OSGi configurations by run mode (author/publish/dev/stage/prod). " +
    "Use OCD annotations with @Designate and keep environment-specific values in " +
    "run-mode folders, not hardcoded in code.",
  cloud_readiness:
    "Remove file system writes, CRX/DE access, install hooks, and replication API usage. " +
    "Use Sling Content Distribution, Asset Compute workers, and cloud-compatible patterns " +
    "per AEMaaCS migration checklist.",
  mutable_content:
    "Move mutable content out of /apps into /conf or /content with proper run-mode " +
    "initialization via repo-init. Only immutable content should be in /apps for " +
    "Cloud Service deployments.",
  dispatcher_cache:
    "Configure dispatcher cache rules with proper TTLs, grace periods, and " +
    "stat-file-level invalidation. Cache HTML with short TTLs; static assets " +
    "with content-hash fingerprints and long TTLs (1 year).",
  clientlib_optimization:
    "Consolidate clientlibs, enable minification and gzip, set proper categories " +
    "and dependencies. Use allowProxy=true for publish delivery and avoid duplicate " +
    "library inclusions across components.",
  query_optimization:
    "Add Oak indexes for all custom queries. Avoid traversal queries (no restrict), " +
    "use query builder with proper predicates, and set limits. Monitor slow query " +
    "logs in production.",
  oak_index:
    "Define custom Oak indexes in /oak:index for all query predicates. Use lucene " +
    "indexes for fulltext, property indexes for exact matches. Include index cost " +
    "estimation and test with explain query.",
  session_save:
    "Avoid session.save() in loops — batch modifications and call save() once. " +
    "Each save triggers observation events and can cause performance degradation " +
    "with large changesets.",
  deprecated_api:
    "Replace deprecated AEM/Sling/JCR APIs with current equivalents. " +
    "WCMUsePojo → Sling Models, @SlingServlet → @Component(service=Servlet.class), " +
    "AdminResourceResolver → ServiceResourceResolver.",
  xss_prevention:
    "Use XSSAPI for context-aware encoding in Java. In HTL, use proper display " +
    "contexts (text, uri, attribute). Never use @{= expression @ context='unsafe'} " +
    "unless content is verified safe.",
  htl_best_practice:
    "Use data-sly-use for Sling Model binding, data-sly-list for iteration, and " +
    "data-sly-test for conditionals. Avoid embedded Java/JSP expressions. Keep HTL " +
    "templates simple with logic in models.",
  content_structure:
    "Follow AEM content hierarchy best practices: /content/site/language-masters " +
    "for i18n, /content/dam for assets, /content/experience-fragments for XF. " +
    "Keep page depth ≤ 5 levels for performance.",
  workflow_optimization:
    "Use transient workflows for asset processing, limit payload size, and add " +
    "timeout handlers. Monitor workflow queue depth and purge completed instances " +
    "on a schedule.",
  replication_pattern:
    "Use Sling Content Distribution (forward distribution) instead of legacy " +
    "replication agents for AEMaaCS. For AMS, configure flush agents with proper " +
    "transport credentials and retry policies.",
  servlet_registration:
    "Register Sling servlets with @Component(service=Servlet.class) and proper " +
    "property annotations (sling.servlet.paths, methods, resourceTypes). Validate " +
    "input, set response content-type, and handle errors gracefully.",
  event_listener:
    "Use Sling Resource Change Listeners or OSGi Event Handlers over JCR observation. " +
    "Filter events narrowly (path, change type), keep handlers lightweight, and " +
    "offload heavy processing to Sling Jobs.",
  sling_job:
    "Use Sling Jobs for asynchronous, guaranteed-delivery processing. Define a " +
    "dedicated topic, configure max retries and retry delay, and add proper " +
    "error handling with job result reporting.",
  test_coverage:
    "Add unit tests (JUnit5 + AEM Mocks/wcm.io), integration tests (AEM Testing " +
    "Clients), and UI tests (Cypress/Selenium). Target 80%+ coverage for custom " +
    "Java code. Gate deployments on test pass.",
  security_general:
    "Apply defense in depth: validate/sanitize all inputs, encode all outputs, " +
    "use service users with least privilege, configure dispatcher deny rules for " +
    "sensitive paths (/crx, /system, /admin).",
  ssrf_prevention:
    "Validate and whitelist all URLs used in server-side HTTP requests. Block " +
    "internal/private IP ranges, use connection timeouts, and avoid user-controlled " +
    "redirect targets.",
  dependency_update:
    "Update dependencies to current stable versions. Pin exact versions in POM, " +
    "check for CVEs with OWASP dependency-check or Snyk, and test compatibility " +
    "with target AEM version.",
  code_complexity:
    "Reduce cyclomatic complexity by extracting methods, using early returns, and " +
    "applying design patterns (Strategy, Template Method). Classes >500 lines or " +
    "methods >50 lines should be refactored.",
  accessibility_fix:
    "Fix WCAG 2.1 AA violations: add ARIA labels, ensure keyboard navigation, " +
    "maintain 4.5:1 contrast ratio, provide text alternatives for images, and " +
    "use semantic HTML elements.",
  seo_best_practice:
    "Add canonical URLs, structured data (JSON-LD), meta descriptions, proper " +
    "heading hierarchy (single h1), and ensure mobile-friendly rendering. Generate " +
    "XML sitemaps with proper lastmod dates.",
  dispatcher_security:
    "Block access to sensitive paths: /crx, /system/console, /bin (selective), " +
    "/libs/granite. Enforce HTTPS, set security headers (CSP, X-Frame-Options, " +
    "HSTS), and validate allowed HTTP methods.",
  validate_general:
    "Recommendation direction looks reasonable. Before implementation, validate " +
    "the finding against real execution paths, add a targeted test, and confirm " +
    "the fix does not alter component rendering or authoring experience.",
  performance_general:
    "Profile the affected code path under production-like load. Fix unbounded " +
    "iterations, add caching where appropriate, and verify response times stay " +
    "within SLA thresholds.",
  maintainability_fix:
    "Improve code maintainability: remove dead code, extract shared logic into " +
    "reusable services, apply consistent naming conventions, and add JavaDoc for " +
    "public APIs.",
  core_components:
    "Extend/proxy Adobe Core Components instead of building from scratch. Use " +
    "Sling Resource Merger for customization, add custom policies in templates, " +
    "and keep BEM CSS conventions.",
};

// ─── Status Prefix Map ────────────────────────────────────────────────────────
const STATUS_PREFIXES: Record<string, string> = {
  aligned: "Aligned",
  aligned_prioritize: "Aligned; prioritize",
  optional: "Optional / style alignment",
  low_risk: "Low-risk correction",
  correct_stronger: "Correct with stronger control",
  needs_review: "Needs review / possible false positive",
};

// Rollout suffix for high-effort items
const ROLLOUT_SUFFIX =
  " Plan rollout in small PRs with rollback notes because " +
  "the estimated effort is high.";

// ─── Category Routing Rules ───────────────────────────────────────────────────
interface CategoryRule {
  default: string;
  default_status?: string;
  overrides?: [string, string, string | null][];
}

const CATEGORY_RULES: Record<string, CategoryRule> = {
  Performance: {
    default: "performance_general",
    overrides: [
      ["Unbounded Query", "query_optimization", null],
      ["Traversal", "query_optimization", null],
      ["Missing Oak Index", "oak_index", null],
      ["Oak Index", "oak_index", null],
      ["session.save", "session_save", null],
      ["Session Save in Loop", "session_save", null],
      ["N+1", "query_optimization", null],
      ["Thread Pool", "performance_general", null],
      ["Resource Leak", "resource_resolver", null],
      ["Cache", "dispatcher_cache", null],
    ],
  },
  "Code Quality": {
    default: "code_complexity",
    overrides: [
      ["God Class", "code_complexity", null],
      ["Large File", "code_complexity", null],
      ["Dead Code", "maintainability_fix", null],
      ["printStackTrace", "validate_general", "low_risk"],
      ["Deprecated", "deprecated_api", null],
      ["WCMUsePojo", "sling_model_best_practice", null],
      ["Complex Method", "code_complexity", null],
      ["Nested Depth", "code_complexity", null],
    ],
  },
  Security: {
    default: "security_general",
    overrides: [
      ["Resource Resolver Leak", "resource_resolver", null],
      ["Admin Session", "service_user", null],
      ["Admin Resource Resolver", "service_user", null],
      ["XSS", "xss_prevention", null],
      ["SSRF", "ssrf_prevention", null],
      ["Cross-Site Scripting", "xss_prevention", null],
      ["Path Traversal", "security_general", null],
      ["Hardcoded", "security_general", "needs_review"],
      ["Sensitive", "security_general", null],
      ["CSRF", "security_general", null],
    ],
  },
  SEO: {
    default: "seo_best_practice",
    overrides: [
      ["Canonical", "seo_best_practice", null],
      ["Structured Data", "seo_best_practice", null],
      ["Meta", "seo_best_practice", null],
      ["Sitemap", "seo_best_practice", null],
      ["Heading", "seo_best_practice", null],
      ["Cache-Control", "dispatcher_cache", null],
    ],
  },
  Accessibility: {
    default: "accessibility_fix",
    overrides: [
      ["ARIA", "accessibility_fix", null],
      ["Contrast", "accessibility_fix", null],
      ["Keyboard", "accessibility_fix", null],
      ["Tab Index", "accessibility_fix", null],
      ["Alt Text", "accessibility_fix", null],
      ["viewport", "accessibility_fix", null],
      ["role=", "accessibility_fix", null],
    ],
  },
  Architecture: {
    default: "sling_model_best_practice",
    overrides: [
      ["WCMUsePojo", "sling_model_best_practice", null],
      ["Sling Model", "sling_model_best_practice", null],
      ["Core Component", "core_components", null],
      ["Mutable Content", "mutable_content", null],
      ["Content Structure", "content_structure", null],
      ["Servlet", "servlet_registration", null],
      ["Service User", "service_user", null],
    ],
  },
  "Sling & OSGi": {
    default: "osgi_config_separation",
    overrides: [
      ["Resource Resolver", "resource_resolver", null],
      ["Service User", "service_user", null],
      ["OSGi Config", "osgi_config_separation", null],
      ["Event", "event_listener", null],
      ["Sling Job", "sling_job", null],
      ["Servlet", "servlet_registration", null],
      ["Filter", "validate_general", null],
    ],
  },
  "Cloud Readiness": {
    default: "cloud_readiness",
    overrides: [
      ["File System", "cloud_readiness", null],
      ["Install Hook", "cloud_readiness", null],
      ["Replication API", "replication_pattern", null],
      ["CRX/DE", "cloud_readiness", null],
      ["Mutable", "mutable_content", null],
      ["Content Distribution", "replication_pattern", null],
      ["Workflow", "workflow_optimization", null],
    ],
  },
  Dispatcher: {
    default: "dispatcher_cache",
    overrides: [
      ["Security", "dispatcher_security", null],
      ["Deny", "dispatcher_security", null],
      ["Cache", "dispatcher_cache", null],
      ["TTL", "dispatcher_cache", null],
      ["Header", "dispatcher_security", null],
      ["Flush", "dispatcher_cache", null],
    ],
  },
  "HTL & Frontend": {
    default: "htl_best_practice",
    overrides: [
      ["HTL", "htl_best_practice", null],
      ["ClientLib", "clientlib_optimization", null],
      ["clientlib", "clientlib_optimization", null],
      ["JavaScript", "clientlib_optimization", null],
      ["CSS", "clientlib_optimization", null],
      ["XSS", "xss_prevention", null],
      ["unsafe", "xss_prevention", null],
    ],
  },
  "Test Coverage": {
    default: "test_coverage",
    overrides: [
      ["JaCoCo", "test_coverage", null],
      ["Unit Test", "test_coverage", null],
      ["Integration Test", "test_coverage", null],
      ["UI Test", "test_coverage", null],
      ["SonarQube", "test_coverage", null],
      ["Functional Test", "test_coverage", null],
    ],
  },
  Maintainability: {
    default: "maintainability_fix",
    overrides: [
      ["Dead Code", "maintainability_fix", null],
      ["Complexity", "code_complexity", null],
      ["Naming", "maintainability_fix", "low_risk"],
      ["Documentation", "maintainability_fix", "optional"],
      ["Duplicate", "maintainability_fix", null],
    ],
  },
  "Dependencies & Versions": {
    default: "dependency_update",
    overrides: [
      ["CVE", "dependency_update", null],
      ["Outdated", "dependency_update", null],
      ["Deprecated", "deprecated_api", null],
      ["Version Conflict", "dependency_update", null],
    ],
  },
};

// ─── Expert Recommendation Generator ──────────────────────────────────────────

function formatRecommendation(
  templateId: string,
  severity: string,
  effort: string,
  statusOverride?: string | null
): string {
  const baseText = TEMPLATES[templateId] ?? TEMPLATES.validate_general;

  let prefix: string;
  if (statusOverride) {
    prefix = STATUS_PREFIXES[statusOverride] ?? "Aligned";
  } else if (severity === "CRITICAL" || severity === "HIGH") {
    prefix = "Aligned; prioritize";
  } else {
    prefix = "Aligned";
  }

  let result = `${prefix}: ${baseText}`;

  if (effort === "High" || effort === "Very High") {
    result += ROLLOUT_SUFFIX;
  }

  return result;
}

export function getExpertRecommendation(
  category: string,
  issueType: string,
  severity = "MEDIUM",
  effort = "Medium"
): string {
  const rules = CATEGORY_RULES[category];
  if (!rules) {
    return formatRecommendation("validate_general", severity, effort);
  }

  let templateId = rules.default;
  let statusOverride: string | null = rules.default_status ?? null;

  for (const override of rules.overrides ?? []) {
    const [keyword, tmplId, sOverride] = override;
    if (issueType.toLowerCase().includes(keyword.toLowerCase())) {
      templateId = tmplId;
      statusOverride = sOverride;
      break;
    }
  }

  return formatRecommendation(templateId, severity, effort, statusOverride);
}
